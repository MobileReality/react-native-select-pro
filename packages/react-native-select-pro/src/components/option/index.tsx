import React, { ComponentPropsWithRef, forwardRef } from 'react';
import {
    StyleSheet,
    Text,
    TextStyle,
    TouchableOpacity,
    ViewStyle,
} from 'react-native';

import {
    COLORS,
    FONT_SIZE,
    ITEM_HEIGHT,
    PADDING,
} from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers/types/OptionalToRequired';
import type { OptionType } from '../../index';
import type { OptionsList } from '../options-list';

type FromSelectComponentProps = Pick<
    ComponentPropsWithRef<typeof OptionsList>,
    | 'optionSelectedStyle'
    | 'optionStyle'
    | 'optionTextStyle'
    | 'onSelect'
    | 'onPressOption'
    | 'OptionComponent'
>;

export type OptionProps = OptionalToRequired<FromSelectComponentProps> & {
    isSelected: boolean;
    option: OptionType;
    optionIndex: number;
};

export type OnChooseOption = () => void;

export const Option = forwardRef<TouchableOpacity, OptionProps>(
    (
        {
            optionSelectedStyle,
            optionStyle,
            optionTextStyle,
            isSelected,
            onPressOption,
            option,
            onSelect,
            OptionComponent,
            optionIndex,
        },
        ref,
    ) => {
        const { label } = option;

        const onChooseOption: OnChooseOption = () => {
            onPressOption(option, optionIndex);
            if (onSelect) {
                onSelect(option, optionIndex);
            }
        };

        if (OptionComponent) {
            return (
                <OptionComponent
                    isSelected={isSelected}
                    option={option}
                    onPressOption={onChooseOption}
                />
            );
        }

        return (
            <TouchableOpacity
                ref={ref}
                accessibilityLabel={`Choose ${label} option`}
                accessibilityRole="button"
                accessible={true}
                style={[
                    styles.option,
                    optionStyle,
                    isSelected && [styles.selected, optionSelectedStyle],
                ]}
                onPress={onChooseOption}
            >
                <Text numberOfLines={1} style={[styles.text, optionTextStyle]}>
                    {label}
                </Text>
            </TouchableOpacity>
        );
    },
);

type Styles = {
    option: ViewStyle;
    selected: ViewStyle;
    text: TextStyle;
};

const styles = StyleSheet.create<Styles>({
    option: {
        height: ITEM_HEIGHT,
        justifyContent: 'center',
        paddingHorizontal: PADDING,
    },
    text: {
        fontSize: FONT_SIZE,
        color: COLORS.BLACK,
        textAlign: 'left',
    },
    selected: {
        backgroundColor: COLORS.SELECTED,
    },
});

Option.displayName = 'Option';
