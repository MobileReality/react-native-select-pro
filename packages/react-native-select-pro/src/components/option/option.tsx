import type { ForwardedRef } from 'react';
import React, { forwardRef, memo } from 'react';
import isEqual from 'react-fast-compare';
import type { TextStyle, View, ViewStyle } from 'react-native';
import { Pressable, StyleSheet, Text } from 'react-native';

import { COLORS, FONT_SIZE, ITEM_HEIGHT, PADDING, PRESSED_STYLE } from '../../constants';
import type { OnChooseOption } from '../../types';

import type { OptionProps } from './option.types';

const OptionComponent = forwardRef(
    <T,>(
        {
            isSelected,
            option,
            optionIndex,
            overrideWithDisabledStyle,
            onPressOption,
            optionButtonProps,
            optionTextProps,
            optionCustomStyles,
            isDisabled,
        }: OptionProps<T>,
        ref: ForwardedRef<View>,
    ) => {
        const onChooseOption: OnChooseOption = () => {
            onPressOption(option, optionIndex);
        };

        const { label } = option;

        return (
            <Pressable
                accessibilityLabel={`Select ${label}`}
                {...optionButtonProps}
                ref={ref}
                accessibilityRole="menuitem"
                accessibilityState={{ disabled: isDisabled }}
                disabled={isDisabled}
                style={({ pressed }) => [
                    styles.option,
                    optionCustomStyles?.container,
                    isSelected && [styles.selected, optionCustomStyles?.selected?.container],
                    pressed && (optionCustomStyles?.pressed ?? PRESSED_STYLE),
                    overrideWithDisabledStyle && styles.disabled,
                ]}
                onPress={onChooseOption}
            >
                <Text
                    numberOfLines={1}
                    {...optionTextProps}
                    style={[
                        styles.text,
                        optionCustomStyles?.text,
                        isSelected && optionCustomStyles?.selected?.text,
                    ]}
                >
                    {label}
                </Text>
            </Pressable>
        );
    },
);

type Styles = {
    option: ViewStyle;
    selected: ViewStyle;
    text: TextStyle;
    disabled: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    option: {
        height: ITEM_HEIGHT,
        justifyContent: 'center',
        paddingHorizontal: PADDING,
    },
    text: {
        fontSize: FONT_SIZE,
        color: COLORS.BLACK,
        textAlign: 'left',
    },
    selected: {
        backgroundColor: COLORS.SELECTED,
    },
    disabled: {
        backgroundColor: COLORS.DISABLED,
    },
});

OptionComponent.displayName = 'OptionComponent';

export const Option = memo(OptionComponent, isEqual);
