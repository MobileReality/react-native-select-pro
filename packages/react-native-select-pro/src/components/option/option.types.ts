import type { OptionalToRequired, OptionStyles, OptionType, SelectProps } from '../../types';
import type { OnPressOptionType } from '../../types';

export type OptionProps<T> = {
    isSelected: boolean;
    isDisabled: boolean;
    overrideWithDisabledStyle: boolean;
    option: OptionType<T>;
    optionIndex: number;
    onPressOption: OnPressOptionType<T>;
    optionCustomStyles: OptionStyles | undefined;
} & OptionalToRequired<Pick<SelectProps<T>, 'optionButtonProps' | 'optionTextProps'>>;
