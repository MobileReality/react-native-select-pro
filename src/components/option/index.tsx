import React, { ComponentProps } from 'react';
import { StyleSheet, Text, TextStyle, TouchableOpacity, ViewStyle } from 'react-native';

import { COLORS, FONT_SIZE, PADDING } from '../../constants/styles';
import type { OptionalToRequired } from '../../helpers';
import type { OptionType } from '../../index';
import type { OptionsList } from '../options-list';

type FromSelectComponentProps = Pick<
    ComponentProps<typeof OptionsList>,
    | 'optionSelectedStyle'
    | 'optionStyle'
    | 'optionTextStyle'
    | 'onSelect'
    | 'onPressOption'
    | 'OptionComponent'
>;

export type OptionProps = OptionalToRequired<FromSelectComponentProps> & {
    isSelected: boolean;
    option: OptionType;
};

export type OnChooseOption = () => void;

export const Option = ({
    optionSelectedStyle,
    optionStyle,
    optionTextStyle,
    isSelected,
    onPressOption,
    option,
    onSelect,
    OptionComponent,
}: OptionProps) => {
    const { label } = option;

    const onChooseOption: OnChooseOption = () => {
        onPressOption(option);
        if (onSelect) {
            onSelect(option);
        }
    };

    if (OptionComponent) {
        return (
            <OptionComponent
                isSelected={isSelected}
                onPressOption={onChooseOption}
                option={option}
            />
        );
    }

    return (
        <TouchableOpacity
            accessibilityLabel={`Choose ${label} option`}
            onPress={onChooseOption}
            style={[
                styles.option,
                optionStyle,
                isSelected && [styles.selected, optionSelectedStyle],
            ]}>
            <Text numberOfLines={1} style={[styles.text, optionTextStyle]}>
                {label}
            </Text>
        </TouchableOpacity>
    );
};

type Styles = {
    option: ViewStyle;
    selected: ViewStyle;
    text: TextStyle;
};

const styles = StyleSheet.create<Styles>({
    option: {
        padding: PADDING,
    },
    text: {
        fontSize: FONT_SIZE,
        color: COLORS.BLACK,
    },
    selected: {
        backgroundColor: COLORS.SELECTED,
    },
});
