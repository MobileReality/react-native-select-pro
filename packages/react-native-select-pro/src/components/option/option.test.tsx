import React from 'react';

import { Option } from './option';

describe('Option', () => {
    it('should generate Option snapshot', () => {
        const option = (
            <Option
                isSelected={false}
                option={{ label: 'test', value: 'test' }}
                optionIndex={0}
                optionSelectedStyle={{}}
                optionStyle={{}}
                optionTextStyle={{}}
                OptionComponent={undefined}
                onPressOption={() => {}}
                onSelect={() => {}}
            />
        );
        expect(option).toMatchSnapshot();
    });
});
