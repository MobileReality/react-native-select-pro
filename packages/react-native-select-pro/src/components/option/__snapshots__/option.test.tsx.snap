// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option should generate Option snapshot 1`] = `
<Option
  isSelected={false}
  onPressOption={[Function]}
  onSelect={[Function]}
  option={
    {
      "label": "test",
      "value": "test",
    }
  }
  optionIndex={0}
  optionSelectedStyle={{}}
  optionStyle={{}}
  optionTextStyle={{}}
/>
`;
